File: spikes/classifier/java/Judge.java


import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * ADR 0020 §D4 — the composable surface: the judge looks <b>on</b> this, <b>asks</b> these,
 * <b>rules</b> by this. Adapters return EXISTING types; here {@link Guardrail}.
 */
public record Judge(Function<Object, Object> on, Map<String, Classifier.Question> ask, Rule rule, boolean failClosed) {

    /** The port's {@code Loop.Guardrail} shape: {@code ""}/{@code "allow"} permits, any other
     *  string denies with that reason. */
    @FunctionalInterface
    public interface Guardrail { String check(Object event); }

    public record Verdict(String kept, Map<String, Object> evidence, String model, boolean calibrated) {}

    /** How a {@link Classifier.Decision} becomes a verdict. */
    public sealed interface Rule {
        Verdict apply(Classifier.Decision d);

        /** Ordered cut-points on one {@code score} answer: the first band whose {@code upto}
         *  exceeds the score wins; the last band is the fallthrough. */
        record Bands(String key, List<Band> bands) implements Rule {
            @Override public Verdict apply(Classifier.Decision d) {
                var a = d.score(key);
                String kept = bands.getLast().name();
                for (Band b : bands) { if (a.score() < b.upto()) { kept = b.name(); break; } }
                var ev = new LinkedHashMap<String, Object>();
                ev.put(key, a.score());
                ev.put("confidence", a.confidence());
                return new Verdict(kept, ev, d.model(), d.calibrated());
            }
        }

        /** The single highest-probability option of one {@code choice} answer. */
        record One(String key) implements Rule {
            @Override public Verdict apply(Classifier.Decision d) {
                var a = d.choice(key);
                return new Verdict(a.choice(), Map.copyOf(a.probabilities()), d.model(), d.calibrated());
            }
        }
    }

    public record Band(String name, double upto) {}

    public static Judge of(Function<Object, Object> on, Map<String, Classifier.Question> ask, Rule rule) {
        return new Judge(on, ask, rule, true);
    }

    public Verdict rule(Classifier c, Object event) {
        try {
            return rule.apply(c.evaluate(on.apply(event), ask));
        } catch (RuntimeException e) {
            if (failClosed) return new Verdict("deny", Map.of("error", e.getMessage()), "", false);
            throw e;
        }
    }

    /**
     * §D4 adapter. INVARIANT: a judge may move a call toward ask/deny, never toward allow — so
     * it returns only "" (allow) or a reason, and is composed BEHIND any earlier guardrail by
     * {@link #firstDenyWins}.
     */
    public Guardrail asGuardrail(Classifier c) {
        return event -> {
            Verdict v = rule(c, event);
            return switch (v.kept()) {
                case "allow" -> "";
                case "ask" -> "ask: " + v.evidence();   // §10 Pending in the real seam
                default -> v.kept() + ": " + v.evidence();
            };
        };
    }

    /** Existing behaviour (golang/agents/loop.go:24-47, java Loop.guardedHooks). */
    public static Guardrail firstDenyWins(List<Guardrail> guardrails) {
        return event -> {
            for (Guardrail g : guardrails) {
                String verdict = g.check(event);
                if (verdict != null && !verdict.isEmpty() && !"allow".equals(verdict)) return verdict;
            }
            return "";
        };
    }
}

File: spikes/classifier/golang/judge.go
package classifier

import (
	"context"
	"fmt"
	"sort"

	tn "github.com/muthuishere/toolnexus/golang"
	"github.com/muthuishere/toolnexus/golang/agents"
)

// Verdict is what a rule produces. Kept is the surviving item(s) for topK/one;
// for bands it is the single band name ("allow" | "ask" | "deny").
type Verdict struct {
	Kept       []string
	Evidence   map[string]float64
	Model      string
	Calibrated bool
	Reason     string
}

// Band is one ordered risk band: everything with score <= Max is this band.
// The LAST band's Max is ignored (it is the catch-all).
type Band struct {
	Max  float64
	Name string
}

// Rule turns a Decision into a Verdict.
type Rule func(Decision) (Verdict, error)

// Bands rules on ONE score answer against ordered bands.
func Bands(key string, bands []Band) Rule {
	return func(d Decision) (Verdict, error) {
		a, err := d.Score(key)
		if err != nil {
			return Verdict{}, err
		}
		name := bands[len(bands)-1].Name
		for _, b := range bands[:len(bands)-1] {
			if a.Score <= b.Max {
				name = b.Name
				break
			}
		}
		ev := map[string]float64{key: a.Score, key + ".confidence": a.Confidence}
		return Verdict{Kept: []string{name}, Evidence: ev, Model: d.Model, Calibrated: d.Calibrated,
			Reason: fmt.Sprintf("%s=%g (%s)", key, a.Score, levelAt(a, a.Score))}, nil
	}
}

func levelAt(a ScoreAnswer, s float64) string {
	i := int(s + 0.5)
	if v, ok := a.Legend[fmt.Sprint(i)]; ok {
		return v
	}
	return "?"
}

// One rules on a choice answer: the argmax option.
func One(key string) Rule {
	return func(d Decision) (Verdict, error) {
		a, err := d.Choice(key)
		if err != nil {
			return Verdict{}, err
		}
		return Verdict{Kept: []string{a.Choice}, Evidence: a.Probabilities,
			Model: d.Model, Calibrated: d.Calibrated}, nil
	}
}

// TopK rules on a choice answer: the k highest-probability options, ties by name.
func TopK(key string, k int) Rule {
	return func(d Decision) (Verdict, error) {
		a, err := d.Choice(key)
		if err != nil {
			return Verdict{}, err
		}
		names := make([]string, 0, len(a.Probabilities))
		for n := range a.Probabilities {
			names = append(names, n)
		}
		sort.Slice(names, func(i, j int) bool {
			if a.Probabilities[names[i]] != a.Probabilities[names[j]] {
				return a.Probabilities[names[i]] > a.Probabilities[names[j]]
			}
			return names[i] < names[j]
		})
		if k < len(names) {
			names = names[:k]
		}
		return Verdict{Kept: names, Evidence: a.Probabilities, Model: d.Model, Calibrated: d.Calibrated}, nil
	}
}

// AtLeast rules on a noul answer against a threshold.
func AtLeast(key string, p float64) Rule {
	return func(d Decision) (Verdict, error) {
		a, err := d.Noul(key)
		if err != nil {
			return Verdict{}, err
		}
		name := "no"
		if a.Noul >= p {
			name = "yes"
		}
		return Verdict{Kept: []string{name}, Evidence: map[string]float64{key: a.Noul},
			Model: d.Model, Calibrated: d.Calibrated}, nil
	}
}

// Judge is the composable surface: look ON this, ASK these, RULE by this.
type Judge struct {
	// On is what the judge may see — the security posture lives here.
	On func(tn.BeforeToolEvent) any
	// Ask are the pre-declared questions.
	Ask map[string]Question
	// Rule turns the Decision into a Verdict.
	Rule Rule
	// FailClosed decides what a backend error means. false ⇒ allow on error.
	FailClosed bool
}

// Rules runs the judge once.
func (j Judge) Rules(ctx context.Context, c *Classifier, ev tn.BeforeToolEvent) (Verdict, error) {
	d, err := c.Evaluate(ctx, j.On(ev), j.Ask)
	if err != nil {
		return Verdict{}, err
	}
	return j.Rule(d)
}

// AsGuardrail adapts a judge to the SHIPPED agents.Guardrail type — "" allows,
// a reason denies. It can only move a call toward ask/deny: it never returns a
// value that widens an earlier denial (the compiled rail is first-deny-wins and
// this adapter has no "allow" override to hand it).
func (j Judge) AsGuardrail(c *Classifier, onAsk func(tn.BeforeToolEvent, Verdict) string) agents.Guardrail {
	return func(ev tn.BeforeToolEvent) string {
		v, err := j.Rules(context.Background(), c, ev)
		if err != nil {
			if j.FailClosed {
				return "judge unavailable: " + err.Error()
			}
			return "" // fail-open, declared
		}
		switch band(v) {
		case "allow":
			return ""
		case "ask":
			if onAsk != nil {
				return onAsk(ev, v)
			}
			return "needs approval: " + v.Reason
		default:
			return "denied by judge: " + v.Reason
		}
	}
}

func band(v Verdict) string {
	if len(v.Kept) == 0 {
		return "deny"
	}
	return v.Kept[0]
}

// Compile is the spike's copy of golang/agents/loop.go:24-47 first-deny-wins,
// so gate 4 can be proven without reaching into the port's unexported helper.
func Compile(rails ...agents.Guardrail) agents.Guardrail {
	return func(ev tn.BeforeToolEvent) string {
		for _, g := range rails {
			if v := g(ev); v != "" && v != "allow" {
				return v
			}
		}
		return ""
	}
}
